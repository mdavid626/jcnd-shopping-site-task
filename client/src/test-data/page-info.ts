import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
