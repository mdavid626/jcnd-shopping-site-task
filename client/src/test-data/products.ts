import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  price: 46.36,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  price: 25.36,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
