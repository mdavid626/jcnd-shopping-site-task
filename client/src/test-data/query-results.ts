export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
