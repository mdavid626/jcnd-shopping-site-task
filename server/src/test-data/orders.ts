import { OrderRequestItem } from '../types/order';

export const testOrderRequestItems1: OrderRequestItem[] = [
  { productId: '63ea6485732ff5cf8b36442c', amount: 2 },
  { productId: '63ea6485732ff5cf8b36442d', amount: 4 },
  { productId: '63ea6485732ff5cf8b36442e', amount: 6 },
];
