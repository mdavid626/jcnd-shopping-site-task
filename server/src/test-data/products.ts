import { Product, ProductsQueryResult } from '../types/product';

const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  category: 'VEGETABLES',
  priceInCents: 123,
  inStock: 5,
};

const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  category: 'VEGETABLES',
  priceInCents: 456,
  inStock: 3,
};

export const testProductsQueryResult1: ProductsQueryResult = {
  nodes: [testProduct1, testProduct2],
  pageInfo: {
    page: 1,
    numberOfPages: 5,
  },
};
