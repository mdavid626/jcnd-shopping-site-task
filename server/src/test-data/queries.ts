export const productsQuery1 = {
  operationName: 'GetProductsQuery',
  variables: { category: 'VEGETABLES', page: 0 },
  query:
    'query GetProductsQuery($category: ProductCategory!, $page: NonNegativeInt!) {\n  products(category: $category, page: $page) {\n    nodes {\n      id\n      priceInCents\n      name\n      inStock\n      __typename\n    }\n    pageInfo {\n      page\n      numberOfPages\n      __typename\n    }\n    __typename\n  }\n}',
};

export const placeOrderQuery1 = {
  operationName: 'PlaceOrder',
  variables: {
    items: [
      { productId: '63ea6485732ff5cf8b36442c', amount: 2 },
      { productId: '63ea6485732ff5cf8b36442d', amount: 1 },
    ],
  },
  query:
    'mutation PlaceOrder($items: [OrderItem!]!) {\n  placeOrder(items: $items)\n}',
};

export const emptyPlaceOrderQuery1 = {
  operationName: 'PlaceOrder',
  variables: {
    items: [],
  },
  query:
    'mutation PlaceOrder($items: [OrderItem!]!) {\n  placeOrder(items: $items)\n}',
};
