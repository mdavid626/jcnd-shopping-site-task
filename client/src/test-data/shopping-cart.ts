import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
