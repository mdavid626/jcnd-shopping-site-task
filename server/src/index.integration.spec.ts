import { Express } from 'express';
import createServer from './server';
import { MongoMemoryServer } from 'mongodb-memory-server';
import mongoose, { connect, Types } from 'mongoose';
import MockDate from 'mockdate';
import OrderModel from './models/order';
import request from 'supertest';
import { placeOrderQuery1, productsQuery1 } from './test-data/queries';

describe('[Integration] app', () => {
  let app: Express;
  let mongod: MongoMemoryServer;

  beforeAll(async () => {
    app = await createServer();
    mongod = await MongoMemoryServer.create();
    mongoose.set('strictQuery', false);
    await connect(mongod.getUri());
  });
  afterAll(async () => {
    await mongoose.disconnect();
    await mongod?.stop();
  });
  beforeEach(async () => {
    MockDate.set(new Date('2023-02-15T16:21:28.519Z'));
    await OrderModel.deleteMany({});
  });
  afterEach(() => {
    MockDate.reset;
  });

  it('should return products', async () => {
    const { body } = await request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(productsQuery1)
      .expect(200);
    expect(body).toMatchSnapshot();
  });

  it('should be able to place order', async () => {
    const { body } = await request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(placeOrderQuery1)
      .expect(200);
    const orders = await OrderModel.find({}).lean();
    expect(orders).toHaveLength(1);
    expect(body).toStrictEqual({
      data: { placeOrder: orders[0]._id.toString() },
    });
    expect(orders[0]).toStrictEqual({
      _id: orders[0]._id,
      createdAt: new Date('2023-02-15T16:21:28.519Z'),
      items: [
        {
          amount: 2,
          productId: new Types.ObjectId('63ea6485732ff5cf8b36442c'),
        },
        {
          amount: 1,
          productId: new Types.ObjectId('63ea6485732ff5cf8b36442d'),
        },
      ],
      updatedAt: new Date('2023-02-15T16:21:28.519Z'),
    });
  });
});
