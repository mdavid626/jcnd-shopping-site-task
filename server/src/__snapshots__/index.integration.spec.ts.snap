// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Integration] app should return products 1`] = `
{
  "data": {
    "products": {
      "__typename": "ProductsQueryResult",
      "nodes": [
        {
          "__typename": "Product",
          "id": "63ea6485732ff5cf8b36442c",
          "inStock": 5,
          "name": "Artichoke",
          "priceInCents": 123,
        },
        {
          "__typename": "Product",
          "id": "63ea6485732ff5cf8b36442d",
          "inStock": 3,
          "name": "Asparagus",
          "priceInCents": 456,
        },
        {
          "__typename": "Product",
          "id": "63ea6485732ff5cf8b36442e",
          "inStock": 25,
          "name": "Eggplant",
          "priceInCents": 4631,
        },
        {
          "__typename": "Product",
          "id": "63ea6485732ff5cf8b36442f",
          "inStock": 977,
          "name": "Avocado",
          "priceInCents": 3838,
        },
        {
          "__typename": "Product",
          "id": "63ea6485732ff5cf8b364430",
          "inStock": 985,
          "name": "Beet",
          "priceInCents": 4361,
        },
      ],
      "pageInfo": {
        "__typename": "PageInfo",
        "numberOfPages": 15,
        "page": 0,
      },
    },
  },
}
`;
